### ISSUE TITLE
Unable to build on ARM: npm WARN old lockfile ui: No matching version found for ui@1.1.0.

### ISSUE THREAD

**Issue by Marietto2008** [getferdi/ferdi#1563]
Hello to everyone.

I've tried to install ferdi on my jetson nano (where it runs ubuntu 21.04 for arm64 for tegra platform). This is what happened...

root@zi-desktop:~/Desktop/Work/I9/Virt# git clone https://github.com/getferdi/ferdi.git

Cloning into 'ferdi'...
remote: Enumerating objects: 30112, done.
remote: Counting objects: 100% (904/904), done.
remote: Compressing objects: 100% (407/407), done.
remote: Total 30112 (delta 516), reused 752 (delta 446), pack-reused 29208
Receiving objects: 100% (30112/30112), 40.17 MiB | 9.49 MiB/s, done.
Resolving deltas: 100% (21871/21871), done.
Updating files: 100% (687/687), done.

root@zi-desktop:~/Desktop/Work/I9/Virt# cd ferdi

root@zi-desktop:~/Desktop/Work/I9/Virt/ferdi# git submodule update --init --recursive

Submodule 'recipes' (https://github.com/getferdi/recipes.git) registered for path 'recipes'
Submodule 'src/internal-server' (https://github.com/getferdi/internal-server.git) registered for path 'src/internal-server'
Cloning into '/home/zi/Desktop/Work/I9/Virt/ferdi/recipes'...
Cloning into '/home/zi/Desktop/Work/I9/Virt/ferdi/src/internal-server'...
Submodule path 'recipes': checked out '8a0971ff9f41fdaa1af15e88740ce9479e1ec6ea'
Submodule path 'src/internal-server': checked out 'b8e39c293c4d58c42f3cddafbd0125fcba5e933e'

root@zi-desktop:~/Desktop/Work/I9/Virt/ferdi# export ELECTRON_CACHE=$HOME/.cache/electron

root@zi-desktop:~/Desktop/Work/I9/Virt/ferdi# export ELECTRON_BUILDER_CACHE=$HOME/.cache/electron-builder

root@zi-desktop:~/Desktop/Work/I9/Virt/ferdi# npx lerna bootstrap

lerna notice cli v4.0.0
lerna info versioning independent
lerna info bootstrap root only
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 
npm WARN old lockfile theme: No matching version found for theme@1.0.14.
npm WARN old lockfile     at module.exports (/usr/local/lib/node_modules/npm/node_modules/npm-pick-manifest/index.js:209:23)
npm WARN old lockfile     at /usr/local/lib/node_modules/npm/node_modules/pacote/lib/registry.js:118:26
npm WARN old lockfile     at async Array.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:700:9)
npm WARN old lockfile  Could not fetch metadata for theme@1.0.14 theme: No matching version found for theme@1.0.14.
npm WARN old lockfile     at module.exports (/usr/local/lib/node_modules/npm/node_modules/npm-pick-manifest/index.js:209:23)
npm WARN old lockfile     at /usr/local/lib/node_modules/npm/node_modules/pacote/lib/registry.js:118:26
npm WARN old lockfile     at async Array.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:700:9) {
npm WARN old lockfile   code: 'ETARGET',
npm WARN old lockfile   type: 'version',
npm WARN old lockfile   wanted: '1.0.14',
npm WARN old lockfile   versions: [
npm WARN old lockfile     '0.0.0', '0.0.1',
npm WARN old lockfile     '0.0.2', '0.0.3',
npm WARN old lockfile     '0.0.4', '0.0.5',
npm WARN old lockfile     '0.0.6', '0.0.7',
npm WARN old lockfile     '0.0.8', '0.1.0'
npm WARN old lockfile   ],
npm WARN old lockfile   distTags: { latest: '0.1.0' },
npm WARN old lockfile   defaultTag: 'latest'
npm WARN old lockfile }
npm WARN old lockfile ui: No matching version found for ui@1.1.0.
npm WARN old lockfile     at module.exports (/usr/local/lib/node_modules/npm/node_modules/npm-pick-manifest/index.js:209:23)
npm WARN old lockfile     at /usr/local/lib/node_modules/npm/node_modules/pacote/lib/registry.js:118:26
npm WARN old lockfile     at async Array.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:700:9)
npm WARN old lockfile  Could not fetch metadata for ui@1.1.0 ui: No matching version found for ui@1.1.0.
npm WARN old lockfile     at module.exports (/usr/local/lib/node_modules/npm/node_modules/npm-pick-manifest/index.js:209:23)
npm WARN old lockfile     at /usr/local/lib/node_modules/npm/node_modules/pacote/lib/registry.js:118:26
npm WARN old lockfile     at async Array.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:700:9) {
npm WARN old lockfile   code: 'ETARGET',
npm WARN old lockfile   type: 'version',
npm WARN old lockfile   wanted: '1.1.0',
npm WARN old lockfile   versions: [
npm WARN old lockfile     '0.0.0', '0.1.0',
npm WARN old lockfile     '0.2.0', '0.2.1',
npm WARN old lockfile     '0.2.2', '0.2.3',
npm WARN old lockfile     '0.2.4'
npm WARN old lockfile   ],
npm WARN old lockfile   distTags: { latest: '0.2.4' },
npm WARN old lockfile   defaultTag: 'latest'
npm WARN old lockfile }
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! Found: react-dom@16.14.0
npm ERR! node_modules/react-dom
npm ERR!   react-dom@"^16.14.0" from the root project
npm ERR!   peer react-dom@"^0.14.0 || ^15.0.0 || ^16" from react-modal@3.11.1
npm ERR!   node_modules/react-modal
npm ERR!     react-modal@"3.11.1" from the root project
npm ERR!   5 more (react-tooltip, react-transition-group, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react-dom@"16.7.0" from @meetfranz/ui@1.1.0
npm ERR! packages/ui
npm ERR!   @meetfranz/ui@1.1.0
npm ERR!   node_modules/@meetfranz/ui
npm ERR!     @meetfranz/ui@"file:packages/ui" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /root/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-06-24T22_47_30_776Z-debug.log
lerna ERR! npm install exited 1 in 'ferdi'
lerna ERR! npm install exited 1 in 'ferdi'
npm ERR! code 1
npm ERR! path /home/zi/Desktop/Work/I9/Virt/ferdi
npm ERR! command failed
npm ERR! command sh -c lerna "bootstrap"

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-06-24T22_47_31_008Z-debug.log

/root/.npm/_logs/2021-06-24T22_47_31_008Z-debug.log

0 verbose cli [
0 verbose cli   '/usr/bin/node',
0 verbose cli   '/usr/share/nodejs/npm/bin/npm-cli.js',
0 verbose cli   'exec',
0 verbose cli   '--',
0 verbose cli   'lerna',
0 verbose cli   'bootstrap'
0 verbose cli ]
1 info using npm@7.5.2
2 info using node@v12.21.0
3 timing config:load:defaults Completed in 6ms
4 timing config:load:file:/usr/share/nodejs/npm/npmrc Completed in 6ms
5 timing config:load:builtin Completed in 7ms
6 timing config:load:cli Completed in 6ms
7 timing config:load:env Completed in 1ms
8 timing config:load:file:/home/zi/Desktop/Work/I9/Virt/ferdi/.npmrc Completed in 2ms
9 timing config:load:project Completed in 6ms
10 timing config:load:file:/root/.npmrc Completed in 0ms
11 timing config:load:user Completed in 1ms
12 timing config:load:file:/etc/npmrc Completed in 0ms
13 timing config:load:global Completed in 0ms
14 timing config:load:cafile Completed in 1ms
15 timing config:load:validate Completed in 2ms
16 timing config:load:setUserAgent Completed in 2ms
17 timing config:load:setEnvs Completed in 3ms
18 timing config:load Completed in 35ms
19 verbose npm-session 3e9c4220cb6695ca
20 timing npm:load Completed in 69ms
21 http fetch GET 304 https://registry.npmjs.org/lerna 397ms (from cache)
22 timing arborist:ctor Completed in 3ms
23 timing arborist:ctor Completed in 1ms
24 timing arborist:ctor Completed in 1ms
25 timing command:exec Completed in 158852ms
26 verbose stack Error: command failed
26 verbose stack     at ChildProcess.<anonymous> (/usr/share/nodejs/@npmcli/promise-spawn/index.js:64:27)
26 verbose stack     at ChildProcess.emit (events.js:314:20)
26 verbose stack     at maybeClose (internal/child_process.js:1022:16)
26 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5)
27 verbose pkgid ferdi@5.6.0-nightly.40
28 verbose cwd /home/zi/Desktop/Work/I9/Virt/ferdi
29 verbose Linux 4.9.140
30 verbose argv "/usr/bin/node" "/usr/share/nodejs/npm/bin/npm-cli.js" "exec" "--" "lerna" "bootstrap"
31 verbose node v12.21.0
32 verbose npm  v7.5.2
33 error code 1
34 error path /home/zi/Desktop/Work/I9/Virt/ferdi
35 error command failed
36 error command sh -c lerna "bootstrap"
37 verbose exit 1

**Comment by Marietto2008** [comment]
@kris7t

**Comment by vraravam** [comment]
```
31 verbose node v12.21.0
32 verbose npm v7.5.2
```
@marcolussetti - looks like you are not using the specific version of `node` and `npm` that Ferdi needs. 

Though I appreciate your intent, I would strongly suggest that you read through the whole `CONTRIBUTING.md` file before reporting tickets. It takes valuable time away from contributors when trying to debug remotely without enough / appropriate data - and then finding that the instructions were not followed.

**Comment by marcolussetti** [comment]
@Marietto2008 see @vraravam 's message above :) (I was tagged instead so wasn't sure if you got the notification)

**Comment by vraravam** [comment]
Sorry @marcolussetti - my bad!

**Comment by mhatvan** [comment]
Best way is to use NVM and install lts/fermium (Node v14) which comes packaged with npm 6.

**Comment by Marietto2008** [comment]
this is what ubuntu 21.04 offers : (and on the nvidia nano,ubuntu is the only os supported by nvidia)

root@zi-desktop:~# node -v
v12.21.0

root@zi-desktop:~# npm -v
7.19.0

root@zi-desktop:~# node-gyp -v
v7.1.2

root@zi-desktop:~# lsb_release -a

No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 21.04
Release:	21.04
Codename:	hirsute

**Comment by vraravam** [comment]
@Marietto2008 - as mentioned above, you will need to either use nvm or asdf to be able to install different versions of node as the project progresses. Please *read* the `CONTRIBUTING.md` file - it clearly lists the versions of `node`, `npm` and `node-gyp` that are needed. Without those *specific versions*, i will not be able to help you.

**Comment by Marietto2008** [comment]
I can try to compile the requested package versions from the source code

**Comment by Marietto2008** [comment]
Anyway,man the instructions aren't clear. Read here :

Currently, these are the combinations of system dependencies that work on an intel-based machines for MacOS/Linux/Windows ---> (building on an ARM-based machine is still a work-in-progress due to node-sass native dependencies) : FROM HERE IT DOES NOT SEEMS THAT THE VERSIONS REQUESTED ARE STILL VALID :

node -v
v14.16.1

npm -v
6.14.12

node-gyp -v
v8.0.0

AND ANYWAY,LET ME UNDERSTAND. rEGARDING NODE,i HAVE 12.21 AND 14.16.1 IS REQUESTED,SO MAYBB i CAN COMPILE THE NEWER VERSIONS FROM THE SOURCE CODE. BUT MY NPM AND MY NODE-GYP ARE NEWER THAT THE REQUESTED VERSIONS. IN THIS CASE SHOULD I DOWNGRADE ?

**Comment by vraravam** [comment]
I would still strongly suggest to use the version managers (nvm, asdf) - so that we are all on the same page when trying to debug and help anything that you notice.

Regarding downgrading the other 2 - im not sure of what the side-effects are. Maybe you can try and get it working? If it works, we can always upgrade based on your learnings.

**Comment by vraravam** [comment]
I have not updated that documentation - thanks for pointing it out. Will push a commit later tonight with that corrected. All these combinations of versions work atm for both x86 and arm64 - mac machines.

**Comment by Marietto2008** [comment]
Do u know if there is a docker image which run Ferdi ? it can help.

**Comment by vraravam** [comment]
Yes - please see the `Dockerfile` already present in the codebase.

**Comment by vraravam** [comment]
The instructions to use docker to build Ferdi are already also present in the `CONTRIBUTING.md` file

**Comment by Marietto2008** [comment]
thanks. I tried,but it didn't work. can u give a look here ? ----> https://github.com/getferdi/ferdi/issues/1567

**Comment by kris7t** [comment]
Unfortunately, the Docker image is only for amd64, not arm64. Contributions for an arm64 Linux build are very much welcome! Unfortunately, we don't have a device to test ourselves.

**Comment by Marietto2008** [comment]
Sp,let's make a recap : running ubuntu 21.04 to run ferdi,I should do something like this ?

node from 12.21.0 --> 14.16.1 : it is an upgrade,but since ubuntu 21.04 is the most updated release,I should recompile node from source ?
npm from 7.19.0--> 6.14.12 : this is a regression,so I should remove npm 7.19.0 from ubuntu 21.04 and I should install the 6.14.12 
node-gyp from 7.1.2 --> 8.0.0 : this is an upgrade : so I should remove 7.1.2 from ubuntu and I should recompile 8 from source. 

is this right ?

**Comment by vraravam** [comment]
@Marietto2008 - true to all that - but, the suggestion from me is to use a package manager (nvm or asdf) to accomplish all of them. That way, your non-ferdi-development experience can remain pristine

**Comment by vraravam** [comment]
@Marietto2008 - coincidentally, I had to setup a new mac, and faced the same issue as you. I found that the npm that was installed by `homebrew` bumped me to `v7.18.x` - and this caused the same exact error that you reported above. Once I removed that version and downgraded to the `6.14.12` as documented in the `CONTRIBUTING.md` file and mentioned above, I am now able to build ferdi from the soruce. All this is on an x86 mac book pro. Once I type in this comment, I will redo the same steps on my daughter's M1 (MBP) and check out if it works for ARM-based machines as well.

**Comment by vraravam** [comment]
Node v14.x is not supported on ARM See [here](https://github.com/nodejs/build/issues/2474#issuecomment-820312698) for the discussion. So, unless we upgrade our node version, we cannot build on ARM machines (for now).

@kris7t - since you are on AUR, can you please confirm if v15 works for you? IF so, I can try the initial builds with v15 and we can get all this merged into the `develop` branch

**Comment by Marietto2008** [comment]
yeah,this is the first try that I did,I have installed junest on my jetson nano / arm64 and then I tried to install ferdi from there. can u give a look if you can understand what's wrong there ?

**Comment by vraravam** [comment]
thanks for the offer @Marietto2008 - but, whatever we do, we don't know if ferdi will build in higher versions of node/npm/node-gyp. I have created a new issue #1569 - which needs to be done before attempting the ARM-builds (imo).

**Comment by Marietto2008** [comment]
anyway man. are u talking about the arm of the apple silicon ? I don't have it. My arm is not the same. I'm on arm64 / tegra. Do u want to use the mac to study the situation ?

**Comment by vraravam** [comment]
Yes - im talking about the Mac M1 ARM. I have a M1 laptop - and am trying with that. (btw, im not a javascript dev - so, taking very cautious steps here)

**Comment by Marietto2008** [comment]
It may work on my tegra / arm64. what u want to do ? do u want to help me here ?

**Comment by vraravam** [comment]
Sure - i can try to help remotely. Lets start off with the exact same versions of the system dependencies: `node`, `node-gyp` and `npm`. Once we have that locked, we can try to solve the next error(s).

**Comment by Marietto2008** [comment]
I'm thinking to configure a docker image with base ubuntu 18.04 installing all the required packages or from binary or from source.

**Comment by Marietto2008** [comment]
let me understand a point : it requires node 14.16.1 or  node 14.15.3 ?

**Comment by vraravam** [comment]
Please see the `.nvmrc` or `package.json` or `CONTRIBUTING.ms` - the versions are specified there

**Comment by vraravam** [comment]
Closing this issue - since the effort and the conversation is being kept up in #1585
